### PR TITLE
Add support for cloud provider config on azure

### LINKED ISSUE #84: azure: re-enable cloud-provider
Currently the cloud provider is crashing the kubelet. We need to figure out:

1) What config is it expecting to read?
2) What is the correct recovery path for nil config? AWS has a nil config recovery path in their plugin entry point.

```
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]: I0320 21:59:57.218818    2628 server.go:312] acquiring file lock on "/var/run/lock/kubelet.lock"
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]: I0320 21:59:57.219230    2628 server.go:317] watching for inotify events for: /var/run/lock/kubelet.lock
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]: I0320 21:59:57.219609    2628 feature_gate.go:189] feature gates: map[]
[Unit]
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]: panic: runtime error: invalid memory address or nil pointer dereference [recovered]
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]:         panic: runtime error: invalid memory address or nil pointer dereference
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0xa7a394]
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]: goroutine 1 [running]:
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]: panic(0x2ea92a0, 0xc420014030)
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]:         /usr/local/go/src/runtime/panic.go:500 +0x1a1
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]: io/ioutil.readAll.func1(0xc4208fe698)
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]:         /usr/local/go/src/io/ioutil/ioutil.go:30 +0x13e
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]: panic(0x2ea92a0, 0xc420014030)
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]:         /usr/local/go/src/runtime/panic.go:458 +0x243
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]: bytes.(*Buffer).ReadFrom(0xc4208fe5e8, 0x0, 0x0, 0xc4202a8c00, 0x0, 0x200)
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]:         /usr/local/go/src/bytes/buffer.go:176 +0x134
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]: io/ioutil.readAll(0x0, 0x0, 0x200, 0x0, 0x0, 0x0, 0x0, 0x0)
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]:         /usr/local/go/src/io/ioutil/ioutil.go:33 +0x147
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]: io/ioutil.ReadAll(0x0, 0x0, 0x5c76860, 0x3435360, 0x5c76860, 0x3435e80, 0x0)
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]:         /usr/local/go/src/io/ioutil/ioutil.go:42 +0x3e
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]: k8s.io/kubernetes/pkg/cloudprovider/providers/azure.NewCloud(0x0, 0x0, 0x7ffebead2f65, 0x5, 0xc4200018f0, 0x1)
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/cloudprovider/providers/azure
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]: k8s.io/kubernetes/pkg/cloudprovider.GetCloudProvider(0x7ffebead2f65, 0x5, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/cloudprovider/plugins.go:85 +
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]: k8s.io/kubernetes/pkg/cloudprovider.InitCloudProvider(0x7ffebead2f65, 0x5, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/cloudprovider/plugins.go:111
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]: k8s.io/kubernetes/cmd/kubelet/app.run(0xc4207eb800, 0x0, 0x496ec8, 0xc4208ffa08)
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubelet/app/server.go:362 +0x
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]: k8s.io/kubernetes/cmd/kubelet/app.Run(0xc4207eb800, 0x0, 0xc4208ffa18, 0xc4208ffa28)
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubelet/app/server.go:270 +0x
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]: main.NewKubelet.func1(0xc420c26870, 0xc42019ef00, 0x0, 0xf, 0x0, 0x0)
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/hyperkube/kubelet.go:37 +0x33
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]: main.(*HyperKube).Run(0xc42040ac60, 0xc42000a610, 0xf, 0xf, 0x0, 0x0)
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/hyperkube/hyperkube.go:179 +0
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]: main.(*HyperKube).RunToExit(0xc42040ac60, 0xc42000a600, 0x10, 0x10)
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/hyperkube/hyperkube.go:189 +0
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]: main.main()
Mar 20 21:59:57 tectonic-master-000000 kubelet-wrapper[2628]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/hyperkube/main.go:46 +0x784
Mar 20 21:59:57 tectonic-master-000000 systemd[1]: kubelet.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
```

### PR DESCRIPTION
**Pull request by discordianfish**
This is WIP, I just wanted to share this early for feedback.

This closes #84 and should be fairly self-explanatory. One change in behaviour is that now you have to specify the provider config as tf variables instead of env variables so we can render the cloud provider config which include these.

I'm also not sure if the cloud config should be a config map or secret. It contains secrets but it's called cloud config, so dunno.. Probably better to use a secret.

Will update PR once I've tested everything in Azure.

### CODE REVIEW

File: modules/bootkube/resources/manifests/kube-apiserver.yaml [hunk 1]
```diff
@@ -56,6 +56,7 @@ spec:
         - --oidc-groups-claim=${oidc_groups_claim}
         - --oidc-ca-file=/etc/kubernetes/secrets/ca.crt
         - --cloud-provider=${cloud_provider}
+        - --cloud-config=/etc/kubernetes/cloud/config.json
```
Comment (coreypobrien) [inline modules/bootkube/resources/manifests/kube-apiserver.yaml]: Not all cloud provider configs are json. Azure uses json or yaml, but the Openstack provider uses gcfg.

Comment (discordianfish) [inline modules/bootkube/resources/manifests/kube-apiserver.yaml]: Hrm okay, would it be reasonable to just drop the extension? Alternatively we need to make this a variable too. I'd rather not do that unless we have to due to other clouds complaining if you point them to a empty config, haven't tried that yet.

**Comment by discordianfish** [comment]
I've just tested this change and with my latest fixes it seems to work, but enabling the cloud provider now causes the kubelet to have CNI / networking related errors:
```
Jun 07 12:48:36 tectonic-master-000000 kubelet-wrapper[1411]: I0607 12:48:36.064733    1411 kubelet.go:1752] skipping pod synchronization - [Kubelet failed to get node info: failed to get external ID from cloud provider: instance not found]
Jun 07 12:48:36 tectonic-master-000000 kubelet-wrapper[1411]: E0607 12:48:36.331817    1411 eviction_manager.go:214] eviction manager: unexpected err: failed GetNode: node 'tectonic-master-000000' not found
Jun 07 12:48:36 tectonic-master-000000 kubelet-wrapper[1411]: W0607 12:48:36.821409    1411 cni.go:157] Unable to update cni config: No networks found in /etc/kubernetes/cni/net.d
Jun 07 12:48:36 tectonic-master-000000 kubelet-wrapper[1411]: E0607 12:48:36.821636    1411 kubelet.go:2067] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
```

I'm debugging this further now but any pointers are welcome.

**Comment by s-urbaniak** [comment]
@discordianfish usually this error appears periodically until the `flannel` image and pod have been started. Can you verify both the presence of the flannel image and the running flannel pod?

**Comment by discordianfish** [comment]
@s-urbaniak I'm recreating the cluster right now (which takes loong time due to azure being incredibly slow..) and will try, but IIRC there weren't any containers running. I assumed that error to be fatal and preventing it from starting containers. Will verify in a bit.

File: modules/azure/vnet/outputs.tf [hunk 2]
```diff
@@ -10,6 +10,7 @@ output "worker_subnet" {
   value = "${var.external_vnet_name == "" ?  join(" ", azurerm_subnet.worker_subnet.*.id) : var.external_worker_subnet_id }"
 }
 
+<<<<<<< HEAD
```
Comment (s-urbaniak) [inline modules/azure/vnet/outputs.tf]: this looks like a merge conflict

Comment (discordianfish) [inline modules/azure/vnet/outputs.tf]: Ah right, forgot to push my the fix. Did that now.

**Comment by discordianfish** [comment]
Testing this I ran into issues with security groups which should be fixed by #1013. Going to cherry-pcik the commits from there to test this further.

**Comment by discordianfish** [comment]
@s-urbaniak Ok, with the fixes from #1013, I'm back at the CNI errors and I can confirm that there are no containers running.

**Comment by discordianfish** [comment]
It looks like on the initial `make apply` it didn't create /etc/kubernetes/manifests/ or something deleted it.
On the next `make apply` run it creates it, but the kubelet still doesn't create the pods. Here is some more from the log:
```
Jun 09 11:54:34 tectonic-master-000000 kubelet-wrapper[1335]: W0609 11:54:34.264362    1335 cni.go:157] Unable to update cni config: No networks found in /etc/kubernetes/cni/net.d
Jun 09 11:54:34 tectonic-master-000000 kubelet-wrapper[1335]: E0609 11:54:34.265248    1335 kubelet.go:2067] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
Jun 09 11:54:37 tectonic-master-000000 kubelet-wrapper[1335]: I0609 11:54:37.096781    1335 kubelet_node_status.go:230] Setting node annotation to enable volume controller attach/detach
Jun 09 11:54:37 tectonic-master-000000 kubelet-wrapper[1335]: E0609 11:54:37.115707    1335 kubelet_node_status.go:73] Unable to construct v1.Node object for kubelet: failed to get external ID from cloud provider: instance not found
Jun 09 11:54:37 tectonic-master-000000 kubelet-wrapper[1335]: E0609 11:54:37.506311    1335 reflector.go:190] k8s.io/kubernetes/pkg/kubelet/kubelet.go:390: Failed to list *v1.Node: Get https://XXXX:443/api/v1/nodes?fieldSelector=metadata.name%3Dtectonic-master-000000&resourceVersion=0: dial tcp 13.93.12.164:443: i/o timeout
Jun 09 11:54:37 tectonic-master-000000 kubelet-wrapper[1335]: E0609 11:54:37.512067    1335 reflector.go:190] k8s.io/kubernetes/pkg/kubelet/config/apiserver.go:46: Failed to list *v1.Pod: Get https://XXXX/api/v1/pods?fieldSelector=spec.nodeName%3Dtectonic-master-000000&resourceVersion=0: dial tcp 13.93.12.164:443: i/o timeout
Jun 09 11:54:37 tectonic-master-000000 kubelet-wrapper[1335]: E0609 11:54:37.512785    1335 reflector.go:190] k8s.io/kubernetes/pkg/kubelet/kubelet.go:382: Failed to list *v1.Service: Get https://XXXX:443/api/v1/services?resourceVersion=0: dial tcp 13.93.12.164:443: i/o timeout
Jun 09 11:54:38 tectonic-master-000000 kubelet-wrapper[1335]: I0609 11:54:38.154441    1335 kubelet.go:1752] skipping pod synchronization - [Kubelet failed to get node info: failed to get external ID from cloud provider: instance not found]
Jun 09 11:54:39 tectonic-master-000000 kubelet-wrapper[1335]: W0609 11:54:39.387831    1335 cni.go:157] Unable to update cni config: No networks found in /etc/kubernetes/cni/net.d
Jun 09 11:54:39 tectonic-master-000000 kubelet-wrapper[1335]: E0609 11:54:39.388736    1335 kubelet.go:2067] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
Jun 09 11:54:43 tectonic-master-000000 kubelet-wrapper[1335]: I0609 11:54:43.276042    1335 kubelet.go:1752] skipping pod synchronization - [Kubelet failed to get node info: failed to get external ID from cloud provider: instance not found]
Jun 09 11:54:44 tectonic-master-000000 kubelet-wrapper[1335]: E0609 11:54:44.073597    1335 eviction_manager.go:214] eviction manager: unexpected err: failed GetNode: node 'tectonic-master-000000' not found
Jun 09 11:54:44 tectonic-master-000000 kubelet-wrapper[1335]: I0609 11:54:44.236663    1335 kubelet_node_status.go:230] Setting node annotation to enable volume controller attach/detach
Jun 09 11:54:44 tectonic-master-000000 kubelet-wrapper[1335]: E0609 11:54:44.256510    1335 kubelet_node_status.go:73] Unable to construct v1.Node object for kubelet: failed to get external ID from cloud provider: instance not found
Jun 09 11:54:44 tectonic-master-000000 kubelet-wrapper[1335]: W0609 11:54:44.511331    1335 cni.go:157] Unable to update cni config: No networks found in /etc/kubernetes/cni/net.d
Jun 09 11:54:44 tectonic-master-000000 kubelet-wrapper[1335]: E0609 11:54:44.511590    1335 kubelet.go:2067] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
```

**Comment by coreypobrien** [comment]
I'm not familiar with Azure specifically, but the CNI stuff is expected during the early stages of deployment (until the CNI is configured by flannel I think), and won't affect the bootstrapping process.  I would be curious to know about `Unable to construct v1.Node object for kubelet: failed to get external ID from cloud provider: instance not found`. That means that the cloud provider wasn't able to find an instance in the cloud that matches itself and thus it won't do any more "Kubernetes" work like deploying the static pods for bootstrapping. Does the hostname of the instance match the name of the instance?

**Comment by discordianfish** [comment]
Right, that's what I'm also suspecting right now. The mapping is based on the hostname? Because no, it doesn't match. :) The hostname is `tectonic-master-000000` while the instance name is `my-cluster-name-masters_0`.

**Comment by discordianfish** [comment]
Looks like the problem is #93

**Comment by discordianfish** [comment]
I'm going to rebase this branch on #1043 which is required for this here to work.

**Comment by discordianfish** [comment]
Ok this also requires the fixes from #1013 which I cherry-picked for now. Once that's merged (and squashed I assume), I'll rebase against master again.

**Comment by discordianfish** [comment]
Rebased on master and cherry-picked the updated #1043 but had to add the cloud-config to other components too. Did that, didn't get to test this though.
If someone else is working on tectonic on azure, feel free to pick up the changes. I won't have time for this until somewhere next week.

**Comment by discordianfish** [comment]
Oh #1043 was already merged, nice! So rebased this against master.

**Comment by discordianfish** [comment]
With #1026 merged this appears to work now as expected (after I fixed my location to not contain any spaces).

**Comment by discordianfish** [comment]
@alexsomesan / @s-urbaniak General question, the cloud config is right now a config map - but it contains secrets. Should I make it a secret or even reuse the existing secret containing the TLS stuff?

**Comment by s-urbaniak** [comment]
despite me initially voting for a config-map initally in https://github.com/coreos/tectonic-installer/issues/84#issuecomment-306451436 you are completely right. If cloud-config includes sensitive credentials data, it should be passed as a secret. Thanks for pointing this out!

**Comment by discordianfish** [comment]
Yay, I've converted the config to secret and tested all changes. Looks like everything is working as hoped: Tested both LoadBalancer allocation as well as volume claims and from what I can tell, everything is working. So from my POV this is ready to get merged! :shipit:

**Comment by alekssaul** [comment]
Pulled the PR branch with a Tectonic User and we deployed to Azure. User reported successful testing with Service type Loadbalancer integration.

File: platforms/azure/main.tf [hunk 3]
```diff
@@ -49,6 +56,23 @@ module "etcd" {
...
 }
 
+# Workaround for https://github.com/hashicorp/terraform/issues/4084
+data "null_data_source" "cloud-provider" {
+  inputs = {
+    "cloud"                      = "${var.tectonic_azure_cloud_environment}"
+    "tenantId"                   = "${var.tectonic_azure_tenant_id}"
+    "subscriptionId"             = "${var.tectonic_azure_subscription_id}"
+    "aadClientId"                = "${var.tectonic_azure_client_id}"
+    "aadClientSecret"            = "${var.tectonic_azure_client_secret}"
```
Comment (cehoffman) [inline platforms/azure/main.tf]: The azure subscription, client id, and tenant id can all come from the terraform provider.

See:

* https://github.com/coreos/tectonic-installer/pull/690/files#diff-a220ed1b5bfb31f80a78b7419badeadaR1
* https://github.com/coreos/tectonic-installer/pull/690/files#diff-a220ed1b5bfb31f80a78b7419badeadaR13

This reduces the configuration an end user has to duplicate.

Comment (cehoffman) [inline platforms/azure/main.tf]: I see you avoid the duplication through configuring the terraform provider using this values. I think that is a reasonable alternative.

Comment (metral) [inline platforms/azure/main.tf]: @discordianfish Can you integrate @cehoffman's suggestion of leveraging the `azurerm_client_config` data source for the subscription_id, client_id and tenant_id? This will help reduce the variables users have to provide.

File: platforms/azure/tectonic.tf [hunk 4]
```diff
@@ -63,7 +65,7 @@ module "tectonic" {
 }
 
 resource "null_resource" "tectonic" {
-  depends_on = ["module.tectonic", "module.masters"]
+  depends_on = ["module.etcd", "module.tectonic", "module.masters"]
```
Comment (metral) [inline platforms/azure/tectonic.tf]: Why did etcd become a dependency here? How is this related to the cloud provider config?

File: examples/terraform.tfvars.azure [hunk 5]
```diff
@@ -10,6 +10,17 @@ tectonic_admin_email = ""
 // Note: This field MUST be set manually prior to creating the cluster.
 tectonic_admin_password_hash = ""
 
+// The client ID to use.
+tectonic_azure_client_id = ""
+
+// The client secret to use.
+tectonic_azure_client_secret = ""
```
Comment (metral) [inline examples/terraform.tfvars.azure]: See @alexsomesan's note about embedding the client secret in the TF configuration file: https://github.com/coreos/tectonic-installer/pull/690/files#diff-75a7ba1b6d2b014a605357c4b944b351

You map it into the k8s components as a k8s Secret, but the client secret is still being configured in TF file itself.

File: platforms/azure/tectonic.tf (hunk 4 above)
Comment (discordianfish) [inline platforms/azure/tectonic.tf]: Oh good point: It's actually a fix for a regression. IIRC in #1026 the dependency on etcd was dropped by not referring it's output anymore. That caused terraform to create the bootkube stuff before etcd which failed.

File: platforms/azure/main.tf (hunk 3 above)
Comment (discordianfish) [inline platforms/azure/main.tf]: [x] done

File: examples/terraform.tfvars.azure (hunk 5 above)
Comment (discordianfish) [inline examples/terraform.tfvars.azure]: This can't be read from azurerm_client_config, so it needs to be a variable. I don't see a problem with that either. You can set them via env variables and doing so is required anyway by not giving explicit variables to the provider. I'm using the following env file which sets all these up
```
export ARM_SUBSCRIPTION_ID=xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
export ARM_CLIENT_ID=xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
export ARM_TENANT_ID=xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
export TF_VAR_tectonic_azure_cloud_environment=AZUREPUBLICCLOUD
export TF_VAR_tectonic_azure_client_secret="abcd"
```

**Comment by discordianfish** [comment]
As requested, this is now reading the ids from azurerm_client_config while keeping environment and client_secret a regular variable.

File: Documentation/variables/azure.md [hunk 6]
```diff
@@ -6,6 +6,9 @@ This document gives an overview of variables used in the Azure platform of the T
 
 | Name | Description | Type | Default |
 |------|-------------|:----:|:-----:|
+| tectonic_azure_client_id | The client ID to use. | string | - |
+| tectonic_azure_client_secret | The client secret to use. | string | - |
```
Comment (cehoffman) [inline Documentation/variables/azure.md]: I don't think `tectonic_azure_client_id` is used anymore with the move to `azurerm_client_config`.

File: Documentation/variables/azure.md [hunk 7]
```diff
@@ -28,6 +31,8 @@ This document gives an overview of variables used in the Azure platform of the T
 | tectonic_azure_ssh_key | Name of an Azure ssh key to use joe-sfo | string | - |
 | tectonic_azure_ssh_network_external | (optional) Network (external) to allow SSH access from. Maps to `source_address_prefix` in Azure. Defaults to `*`. Can be external to Azure environment. Allowed values: [network CIDR (i.e., 10.0.0.0/16) | `VirtualNetwork` | `Internet` | `*` ] | string | `*` |
 | tectonic_azure_ssh_network_internal | Network (internal) to allow SSH access from. Maps to `source_address_prefix` in Azure. Defaults to `VirtualNetwork`. Should be internal to Azure environment. Allowed values: [network CIDR (i.e., 10.0.0.0/16) | `VirtualNetwork` | `Internet` | `*` ] | string | `VirtualNetwork` |
+| tectonic_azure_subscription_id | The subscription ID to use. | string | - |
+| tectonic_azure_tenant_id | The tenant ID to use. | string | - |
```
Comment (cehoffman) [inline Documentation/variables/azure.md]: Similar comment to the client_id. `azurerm_client_config` obsoleted these variable inputs.

**Comment by discordianfish** [comment]
@cehoffman good catch, sorry missed running make docs/example. Done!

**Comment by metral** [comment]
@s-urbaniak @alexsomesan How is this PR shaping out? Is there any other feedback needed to get this merged in?

**Comment by cehoffman** [comment]
I think it looks good. Only comment I have is that including the route table configuration property would keep visibility of the configuration operation going forward when integrating #178. At least I think the Azure CNI may integrate with the route table.

**Comment by cehoffman** [comment]
Is there any intent to create a usable dynamic PV provisioner? If so there needs to be a storage account created with a container inside that has the exact name `vhds` in order to the provider to find where PVs should be stored.

**Comment by discordianfish** [comment]
@cehoffman Not sure what you mean? With this PR things work as expected: You need to create a StorageClass like this:

```
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: standard
provisioner: kubernetes.io/azure-disk
```

And then you can use regular persistent volume claims.

Either way, I'd like to get this merged soon and address any further improvement in another PR.

**Comment by cehoffman** [comment]
At least when I did this in 1.6.0 I needed to explicitly create the `vhds` container inside a storage account. Looking at 1.6.4 it appears there is logic to create the container if it does not exist. It also appears that it was there in 1.6.0 too, but I know it wasn't functional in mid May when I did my own cloud config integration. From code it appears it should work, and if you say it is then that is good enough for me at the moment.

**Comment by metral** [comment]
@alexsomesan @s-urbaniak Thoughts on merging this in?

**Comment by alexsomesan** [comment]
Did a test re-basing of this change on the overall azure cleanup branch. No major conflicts.
As discussed with @discordianfish, I'm going to merge this into current master now.

There are a few things that need to be adapted. We try to observe the rule that top-level templates only invoke module, not individual resources. 
In that spirit I suggested moving the top-level null_datasource into a module and passing around rendered cloud-config between modules instead.
Once this is merged, I will do the necessary adaptations as a follow-up in the clean-up PR.

**Comment by alexsomesan** [comment]
retest this please

**Comment by alexsomesan** [comment]
retest this please

**Comment by discordianfish** [comment]
What's with this now? Can't see the build log.

**Comment by alekssaul** [comment]
I believe the issue is, bootkube module has a new variable;
```
variable "cloud_provider_config" { 
description = "Content of cloud provider config"
type        = "string"
}
```
without a default. so our AWS tests are failing because AWS platform calls bootkube without specifying `cloud_provider_config`

File: modules/bootkube/variables.tf [hunk 8]
```diff
@@ -40,6 +40,11 @@ variable "cloud_provider" {
   type        = "string"
 }
 
+variable "cloud_provider_config" {
+  description = "Content of cloud provider config"
+  type        = "string"
+}
+
```
Comment (alekssaul) [inline modules/bootkube/variables.tf]: This may need a default variable

**Comment by discordianfish** [comment]
I've set cloud_provider_config default to "". Can someone paste the full error though? That would help, not sure why it fails if it's not using this on AWS.

**Comment by alexsomesan** [comment]
restest this please

**Comment by discordianfish** [comment]
@alexsomesan Any news?

**Comment by s-urbaniak** [comment]
@discordianfish I am picking this up, sorry for the delay

**Comment by s-urbaniak** [comment]
Closing in favor of #1381
